### PR TITLE
Replacing select and FD_SET with epoll to support more than 1024 fake switches

### PR DESCRIPTION
**Pull request by justin-labry**
I replaced select and FD_SET functions in cbench.c to epoll to support more than 1024 fake switches.
I have tested it with 2048 switches and it works well.
Before run it, you need to set open files limits to more than 1024 using "ulimit -n 4096."
I set the __FD_SETSIZE to 4096 at /usr/include/bits/typesizes.h.
In case you need more than 4096 open connections, you can set the limits with a higher value.
I guess that depends on systems.

### CODE REVIEW

File: cbench/cbench.c [hunk 1]
```diff
@@ -69,17 +74,20 @@ double run_test(int n_fakeswitches, struct fakeswitch * fakeswitches, int mstest
         timersub(&now, &then, &diff);
         if( (1000* diff.tv_sec  + (float)diff.tv_usec/1000)> total_wait)
             break;
-        for(i = 0; i< n_fakeswitches; i++)
-            fakeswitch_set_pollfd(&fakeswitches[i], &pollfds[i]);
 
-        poll(pollfds, n_fakeswitches, 1000);      // block until something is ready or 100ms passes
+	for(i = 0; i < MAX_EVENTS; i++) {
```
Comment (andi-bigswitch) [inline cbench/cbench.c]: Nit - indentation. The cbench code uses 4 chars of space as indentation - your editor seems to use tab -  please reformat to four spaces for consistency.

**Comment by andi-bigswitch** [comment]
Hi Justin, thanks for your pull request. Aside from the whitespace nit above I have one concern with it. Currently cbench is portable and can be built/run on other unixes, e.g., MacOS. With your pull request, it would only work in Linux. I would like to retain the portability, while also enabling Linux users to use >1024 FDs.

Could you rework your PR somewhat to remain portability? For this, you would have to adapt the function signatures to work with either poll or epoll, then use `#ifdef` to chose which one to use.

**Comment by justin-labry** [comment]
Then, I'll try to use the old code for portability (MacOS and other Unix) and the epoll only when the user uses Linux using #ifdef. Is this sound okay?

**Comment by justin-labry** [comment]
I finished the code. Now, Cbench uses the old code for portability (for MacOS and other Unixes) and epoll for Linux users. I tested it using #undef **linux** to check the old code still works.

**Comment by andi-bigswitch** [comment]
Looks like your non-linux codepath doesn't compile quite yet:

```
gcc -DHAVE_CONFIG_H -I. -I..     -g -O2 -Wall -I/Users/andi/dev/oflops/../openflow/include -Wshadow -MT myargs.o -MD -MP -MF .deps/myargs.Tpo -c -o myargs.o myargs.c
mv -f .deps/myargs.Tpo .deps/myargs.Po
gcc -DHAVE_CONFIG_H -I. -I..     -g -O2 -Wall -I/Users/andi/dev/oflops/../openflow/include -Wshadow -MT cbench.o -MD -MP -MF .deps/cbench.Tpo -c -o cbench.o cbench.c
cbench.c:446:111: error: too few arguments to function call, expected 9, have 8
        fakeswitch_init(&fakeswitches[i], sock, 65536, debug, delay, mode, total_mac_addresses, learn_dst_macs);
        ~~~~~~~~~~~~~~~                                                                                    
```

**Comment by andi-bigswitch** [comment]
One other request -- instead of directly checking for `#ifdef __linux__` in various places, could you define a custom macro (e.g., `USE_EPOLL`) that you check for.

Then, have _one_ place (e.g., in `cbench.h`), where you do

```
#ifdef __linux__
#define USE_EPOLL
#endif
```

That way, you can also easily test the non-epoll codepath (even under Linux) :)

**Comment by justin-labry** [comment]
I did exactly what you told me (indirect macro) for more flexibility.
I placed it in cbench.h and tested with my Ubuntu 12.04 machine for both the old and epoll code paths.

**Comment by andi-bigswitch** [comment]
Awesome! Thanks for working on this.

**Pull request merged**